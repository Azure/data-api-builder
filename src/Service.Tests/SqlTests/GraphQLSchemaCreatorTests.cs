using System.Collections.Generic;
using System.Text.Json;
using Azure.DataApiBuilder.Auth;
using Azure.DataApiBuilder.Config;
using Azure.DataApiBuilder.Service.Configurations;
using Azure.DataApiBuilder.Service.Services;
using Azure.DataApiBuilder.Service.Tests.GraphQLBuilder.Helpers;
using HotChocolate;
using Microsoft.Extensions.Logging;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Moq;
using Snapshooter.MSTest;

namespace Azure.DataApiBuilder.Service.Tests.SqlTests
{
    [TestClass]
    public class GraphQLSchemaCreatorTests
    {
        private static DatabaseTable CreateDatabaseTable(string name)
        {
            DatabaseTable databaseTable = new("dbo", name.ToLower())
            {
                TableDefinition = new()
            };
            databaseTable.TableDefinition.Columns.Add("id", new() { IsAutoGenerated = true, SystemType = typeof(int) });
            databaseTable.TableDefinition.Columns.Add("name", new() { SystemType = typeof(string) });
            databaseTable.TableDefinition.PrimaryKey.Add("id");
            return databaseTable;
        }

        private static Entity CreateEntity(string name)
        {
            return new Entity(JsonSerializer.SerializeToElement(name.ToLower()), null, name, new[] { new PermissionSetting(Operation.Read.ToString(), new[] { "*" }), new PermissionSetting(Operation.Create.ToString(), new[] { "*" }) }, null, null);
        }

        [TestMethod]
        public void SingleObjectSchema()
        {
            DataSource dataSource = new(DatabaseType.mssql);
            MsSqlOptions dbOptions = new();

            Dictionary<string, Entity> entities = new()
            {
                {"Author", CreateEntity("Author") }
            };
            RuntimeConfig config = new("", dataSource, null, dbOptions, null, null, null, entities);
            Mock<ILogger<RuntimeConfigProvider>> configProviderLogger = new();
            RuntimeConfigProvider runtimeConfigProvider = new(config, configProviderLogger.Object);
            DatabaseTable databaseTable = CreateDatabaseTable("Author");

            MsSqlMetadataProvider metadataProvider = new(runtimeConfigProvider, null, null, new Mock<ILogger<ISqlMetadataProvider>>().Object)
            {
                EntityToDatabaseObject = new() { { "Author", databaseTable } }
            };

            Dictionary<string, EntityMetadata> entityMap = GraphQLTestHelpers.CreateStubEntityPermissionsMap(
                    entityNames: new[] { "Author" },
                    operations: new Operation[] { Operation.Create, Operation.Read, Operation.Update, Operation.Delete },
                    roles: new string[] { "anonymous", "authenticated" }
                );

            Mock<IAuthorizationResolver> authResolver = new();
            authResolver.Setup(x => x.GetRolesForEntity(It.IsAny<string>())).Returns(new[] { "*" });
            authResolver.Setup(x => x.GetRolesForField(It.IsAny<string>(), It.IsAny<string>(), It.IsAny<Operation>())).Returns(new[] { "*" });
            authResolver.Setup(x => x.EntityPermissionsMap).Returns(entityMap);

            GraphQLSchemaCreator schemaCreator = new(
                runtimeConfigProvider,
                null,
                null,
                metadataProvider,
                authResolver.Object);

            ISchemaBuilder schemaBuilder = schemaCreator.InitializeSchemaAndResolvers(new SchemaBuilder());
            ISchema schema = schemaBuilder.Create();

            Snapshot.Match(schema.ToString());
        }

        [TestMethod]
        public void MultipleObjectSchema()
        {
            DataSource dataSource = new(DatabaseType.mssql);
            MsSqlOptions dbOptions = new();

            Dictionary<string, Entity> entities = new()
            {
                {"Author", CreateEntity("Author") },
                {"Book", CreateEntity("Book") }
            };
            RuntimeConfig config = new("", dataSource, null, dbOptions, null, null, null, entities);
            Mock<ILogger<RuntimeConfigProvider>> configProviderLogger = new();
            RuntimeConfigProvider runtimeConfigProvider = new(config, configProviderLogger.Object);

            MsSqlMetadataProvider metadataProvider = new(runtimeConfigProvider, null, null, new Mock<ILogger<ISqlMetadataProvider>>().Object)
            {
                EntityToDatabaseObject = new() {
                    { "Author", CreateDatabaseTable("Author") },
                    { "Book", CreateDatabaseTable("Book") }
                }
            };

            Dictionary<string, EntityMetadata> entityMap = GraphQLTestHelpers.CreateStubEntityPermissionsMap(
                    entityNames: new[] { "Author", "Book" },
                    operations: new Operation[] { Operation.Create, Operation.Read, Operation.Update, Operation.Delete },
                    roles: new string[] { "anonymous", "authenticated" }
                );

            Mock<IAuthorizationResolver> authResolver = new();
            authResolver.Setup(x => x.GetRolesForEntity(It.IsAny<string>())).Returns(new[] { "*" });
            authResolver.Setup(x => x.GetRolesForField(It.IsAny<string>(), It.IsAny<string>(), It.IsAny<Operation>())).Returns(new[] { "*" });
            authResolver.Setup(x => x.EntityPermissionsMap).Returns(entityMap);

            GraphQLSchemaCreator schemaCreator = new(
                runtimeConfigProvider,
                null,
                null,
                metadataProvider,
                authResolver.Object);

            ISchemaBuilder schemaBuilder = schemaCreator.InitializeSchemaAndResolvers(new SchemaBuilder());
            ISchema schema = schemaBuilder.Create();

            Snapshot.Match(schema.ToString());
        }

        [TestMethod]
        public void OneWayRelationship()
        {
            DataSource dataSource = new(DatabaseType.mssql);
            MsSqlOptions dbOptions = new();

            Entity authorEntity = CreateEntity("Author") with
            {
                Relationships = new() {
                    { "Book", new(Cardinality.Many, "dbo.Book", new[] { "book_id" }, new[] { "id" }, null, null, null) }
                }
            };
            Dictionary<string, Entity> entities = new()
            {
                {"Author", authorEntity },
                {"Book", CreateEntity("Book") }
            };
            RuntimeConfig config = new("", dataSource, null, dbOptions, null, null, null, entities);
            Mock<ILogger<RuntimeConfigProvider>> configProviderLogger = new();
            RuntimeConfigProvider runtimeConfigProvider = new(config, configProviderLogger.Object);

            DatabaseTable authorTable = CreateDatabaseTable("Author");
            authorTable.TableDefinition.Columns.Add("book_id", new ColumnDefinition { SystemType = typeof(int) });
            MsSqlMetadataProvider metadataProvider = new(runtimeConfigProvider, null, null, new Mock<ILogger<ISqlMetadataProvider>>().Object)
            {
                EntityToDatabaseObject = new() {
                    { "Author", authorTable },
                    { "Book", CreateDatabaseTable("Book") }
                }
            };

            Dictionary<string, EntityMetadata> entityMap = GraphQLTestHelpers.CreateStubEntityPermissionsMap(
                    entityNames: new[] { "Author", "Book" },
                    operations: new Operation[] { Operation.Create, Operation.Read, Operation.Update, Operation.Delete },
                    roles: new string[] { "anonymous", "authenticated" }
                );

            Mock<IAuthorizationResolver> authResolver = new();
            authResolver.Setup(x => x.GetRolesForEntity(It.IsAny<string>())).Returns(new[] { "*" });
            authResolver.Setup(x => x.GetRolesForField(It.IsAny<string>(), It.IsAny<string>(), It.IsAny<Operation>())).Returns(new[] { "*" });
            authResolver.Setup(x => x.EntityPermissionsMap).Returns(entityMap);

            GraphQLSchemaCreator schemaCreator = new(
                runtimeConfigProvider,
                null,
                null,
                metadataProvider,
                authResolver.Object);

            ISchemaBuilder schemaBuilder = schemaCreator.InitializeSchemaAndResolvers(new SchemaBuilder());
            ISchema schema = schemaBuilder.Create();

            Snapshot.Match(schema.ToString());
        }

        [TestMethod]
        public void RecursiveRelationship()
        {
            DataSource dataSource = new(DatabaseType.mssql);
            MsSqlOptions dbOptions = new();

            Entity authorEntity = CreateEntity("Author") with
            {
                Relationships = new() {
                    { "Book", new(Cardinality.Many, "dbo.Book", null, null, "dbo.Book_Author", new[] { "id"}, new [] { "author_id" }) }
                }
            };
            Entity bookEntity = CreateEntity("Book") with
            {
                Relationships = new() {
                    { "Author", new(Cardinality.Many, "dbo.Author", null, null, "dbo.Book_Author", new[] { "id"}, new [] { "book_id" }) }
                }
            };
            Dictionary<string, Entity> entities = new()
            {
                {"Author", authorEntity },
                {"Book", bookEntity }
            };
            RuntimeConfig config = new("", dataSource, null, dbOptions, null, null, null, entities);
            Mock<ILogger<RuntimeConfigProvider>> configProviderLogger = new();
            RuntimeConfigProvider runtimeConfigProvider = new(config, configProviderLogger.Object);

            MsSqlMetadataProvider metadataProvider = new(runtimeConfigProvider, null, null, new Mock<ILogger<ISqlMetadataProvider>>().Object)
            {
                EntityToDatabaseObject = new() {
                    { "Author", CreateDatabaseTable("Author") },
                    { "Book", CreateDatabaseTable("Book") }
                }
            };

            Dictionary<string, EntityMetadata> entityMap = GraphQLTestHelpers.CreateStubEntityPermissionsMap(
                    entityNames: new[] { "Author", "Book" },
                    operations: new Operation[] { Operation.Create, Operation.Read, Operation.Update, Operation.Delete },
                    roles: new string[] { "anonymous", "authenticated" }
                );

            Mock<IAuthorizationResolver> authResolver = new();
            authResolver.Setup(x => x.GetRolesForEntity(It.IsAny<string>())).Returns(new[] { "*" });
            authResolver.Setup(x => x.GetRolesForField(It.IsAny<string>(), It.IsAny<string>(), It.IsAny<Operation>())).Returns(new[] { "*" });
            authResolver.Setup(x => x.EntityPermissionsMap).Returns(entityMap);

            GraphQLSchemaCreator schemaCreator = new(
                runtimeConfigProvider,
                null,
                null,
                metadataProvider,
                authResolver.Object);

            ISchemaBuilder schemaBuilder = schemaCreator.InitializeSchemaAndResolvers(new SchemaBuilder());
            ISchema schema = schemaBuilder.Create();

            Snapshot.Match(schema.ToString());
        }
    }
}
