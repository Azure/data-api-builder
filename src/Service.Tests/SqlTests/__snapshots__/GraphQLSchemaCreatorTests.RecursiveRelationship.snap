schema {
  query: Query
  mutation: Mutation
}

type Author @model(name: "Author") @authorize(apply: BEFORE_RESOLVER, roles: [ "*" ]) {
  id: Int! @primaryKey(databaseType: "Int32") @autoGenerated @authorize(apply: BEFORE_RESOLVER, roles: [ "*" ])
  name: String! @authorize(apply: BEFORE_RESOLVER, roles: [ "*" ])
  Book("The number of items to return from the page start point" first: Int "A pagination token from a previous query to continue through a paginated list" after: String "Filter options for query" filter: BookFilterInput "Ordering options for query" orderBy: BookOrderByInput): BookConnection! @relationship(target: "Book", cardinality: "Many")
}

"The return object from a filter query that supports a pagination token for paging through results"
type AuthorConnection {
  "The list of items that matched the filter"
  items: [Author!]!
  "A pagination token to provide to subsequent pages of a query"
  endCursor: String
  "Indicates if there are more pages of items to return"
  hasNextPage: Boolean!
}

type Book @model(name: "Book") @authorize(apply: BEFORE_RESOLVER, roles: [ "*" ]) {
  id: Int! @primaryKey(databaseType: "Int32") @autoGenerated @authorize(apply: BEFORE_RESOLVER, roles: [ "*" ])
  name: String! @authorize(apply: BEFORE_RESOLVER, roles: [ "*" ])
  Author("The number of items to return from the page start point" first: Int "A pagination token from a previous query to continue through a paginated list" after: String "Filter options for query" filter: AuthorFilterInput "Ordering options for query" orderBy: AuthorOrderByInput): AuthorConnection! @relationship(target: "Author", cardinality: "Many")
}

"The return object from a filter query that supports a pagination token for paging through results"
type BookConnection {
  "The list of items that matched the filter"
  items: [Book!]!
  "A pagination token to provide to subsequent pages of a query"
  endCursor: String
  "Indicates if there are more pages of items to return"
  hasNextPage: Boolean!
}

type Mutation {
  "Creates a new Author"
  createAuthor("Input representing all the fields for creating Author" item: CreateAuthorInput!): Author
  "Updates a Author"
  updateAuthor("The ID of the item being updated." id: Int! "Input representing all the fields for updating Author" item: UpdateAuthorInput!): Author
  "Delete a Author"
  deleteAuthor("The ID of the item being deleted." id: Int!): Author
  "Creates a new Book"
  createBook("Input representing all the fields for creating Book" item: CreateBookInput!): Book
  "Updates a Book"
  updateBook("The ID of the item being updated." id: Int! "Input representing all the fields for updating Book" item: UpdateBookInput!): Book
  "Delete a Book"
  deleteBook("The ID of the item being deleted." id: Int!): Book
}

type Query {
  "Get a list of all the Author items from the database"
  authors("The number of items to return from the page start point" first: Int "A pagination token from a previous query to continue through a paginated list" after: String "Filter options for query" filter: AuthorFilterInput "Ordering options for query" orderBy: AuthorOrderByInput): AuthorConnection!
  "Get a Author from the database by its ID\/primary key"
  author_by_pk(id: Int!): Author
  "Get a list of all the Book items from the database"
  books("The number of items to return from the page start point" first: Int "A pagination token from a previous query to continue through a paginated list" after: String "Filter options for query" filter: BookFilterInput "Ordering options for query" orderBy: BookOrderByInput): BookConnection!
  "Get a Book from the database by its ID\/primary key"
  book_by_pk(id: Int!): Book
}

"Filter input for Author GraphQL type"
input AuthorFilterInput {
  "Filter options for id"
  id: IntFilterInput
  "Filter options for name"
  name: StringFilterInput
  "Filter options for Book"
  Book: BookFilterInput
  "Conditions to be treated as AND operations"
  and: [AuthorFilterInput]
  "Conditions to be treated as OR operations"
  or: [AuthorFilterInput]
}

"Order by input for Author GraphQL type"
input AuthorOrderByInput {
  "Order by options for id"
  id: OrderBy
  "Order by options for name"
  name: OrderBy
  "Order by options for Book"
  Book: BookOrderByInput
  "Conditions to be treated as AND operations"
  and: [AuthorOrderByInput]
  "Conditions to be treated as OR operations"
  or: [AuthorOrderByInput]
}

"Filter input for Book GraphQL type"
input BookFilterInput {
  "Filter options for id"
  id: IntFilterInput
  "Filter options for name"
  name: StringFilterInput
  "Filter options for Author"
  Author: AuthorFilterInput
  "Conditions to be treated as AND operations"
  and: [BookFilterInput]
  "Conditions to be treated as OR operations"
  or: [BookFilterInput]
}

"Order by input for Book GraphQL type"
input BookOrderByInput {
  "Order by options for id"
  id: OrderBy
  "Order by options for name"
  name: OrderBy
  "Order by options for Author"
  Author: AuthorOrderByInput
  "Conditions to be treated as AND operations"
  and: [BookOrderByInput]
  "Conditions to be treated as OR operations"
  or: [BookOrderByInput]
}

"Input type for creating Author"
input CreateAuthorInput {
  "Input for field name on type CreateAuthorInput"
  name: String!
}

"Input type for creating Book"
input CreateBookInput {
  "Input for field name on type CreateBookInput"
  name: String!
}

input DefaultValue @oneOf {
  Byte: Byte
  Short: Short
  Int: Int
  Long: Long
  String: String
  Boolean: Boolean
  Single: Single
  Float: Float
  Decimal: Decimal
  DateTime: DateTime
  ByteArray: ByteArray
}

"Input type for adding Int filters"
input IntFilterInput {
  "Equals"
  eq: Int
  "Greater Than"
  gt: Int
  "Greater Than or Equal To"
  gte: Int
  "Less Than"
  lt: Int
  "Less Than or Equal To"
  lte: Int
  "Not Equals"
  neq: Int
  "Not null test"
  isNull: Boolean
}

"Input type for adding String filters"
input StringFilterInput {
  "Equals"
  eq: String
  "Contains"
  contains: String
  "Not Contains"
  notContains: String
  "Starts With"
  startsWith: String
  "Ends With"
  endsWith: String
  "Not Equals"
  neq: String
  "Case Insensitive"
  caseInsensitive: Boolean = false
  "Not null test"
  isNull: Boolean
}

"Input type for updating Author"
input UpdateAuthorInput {
  "Input for field name on type UpdateAuthorInput"
  name: String
}

"Input type for updating Book"
input UpdateBookInput {
  "Input for field name on type UpdateBookInput"
  name: String
}

enum ApplyPolicy {
  BEFORE_RESOLVER
  AFTER_RESOLVER
}

enum OrderBy {
  ASC
  DESC
}

directive @authorize("The name of the authorization policy that determines access to the annotated resource." policy: String "Roles that are allowed to access the annotated resource." roles: [String!] "Defines when when the resolver shall be executed.By default the resolver is executed after the policy has determined that the current user is allowed to access the field." apply: ApplyPolicy! = BEFORE_RESOLVER) repeatable on SCHEMA | OBJECT | FIELD_DEFINITION

"Indicates that a field is auto generated by the database."
directive @autoGenerated on FIELD_DEFINITION

"The default value to be used when creating an item."
directive @defaultValue(value: DefaultValue) on FIELD_DEFINITION

"The `@defer` directive may be provided for fragment spreads and inline fragments to inform the executor to delay the execution of the current fragment to indicate deprioritization of the current fragment. A query with `@defer` directive will cause the request to potentially return multiple responses, where non-deferred data is delivered in the initial response and data deferred is delivered in a subsequent response. `@include` and `@skip` take precedence over `@defer`."
directive @defer("If this argument label has a value other than null, it will be passed on to the result of this defer directive. This label is intended to give client applications a way to identify to which fragment a deferred result belongs to." label: String "Deferred when true." if: Boolean) on FRAGMENT_SPREAD | INLINE_FRAGMENT

"A directive to indicate the type maps to a storable entity not a nested entity."
directive @model("Underlying name of the database entity." name: String) on OBJECT

"The `@oneOf` directive is used within the type system definition language\n to indicate:\n\n - an Input Object is a Oneof Input Object, or\n - an Object Type's Field is a Oneof Field."
directive @oneOf on INPUT_OBJECT

"A directive to indicate the primary key field of an item."
directive @primaryKey("The underlying database type." databaseType: String) on FIELD_DEFINITION

"A directive to indicate the relationship between two tables"
directive @relationship("The name of the entity the relationship targets" target: String "The relationship cardinality" cardinality: String) on FIELD_DEFINITION

"The `@specifiedBy` directive is used within the type system definition language to provide a URL for specifying the behavior of custom scalar definitions."
directive @specifiedBy("The specifiedBy URL points to a human-readable specification. This field will only read a result for scalar types." url: String!) on SCALAR

"The `@stream` directive may be provided for a field of `List` type so that the backend can leverage technology such as asynchronous iterators to provide a partial list in the initial response, and additional list items in subsequent responses. `@include` and `@skip` take precedence over `@stream`."
directive @stream("If this argument label has a value other than null, it will be passed on to the result of this stream directive. This label is intended to give client applications a way to identify to which fragment a streamed result belongs to." label: String "The initial elements that shall be send down to the consumer." initialCount: Int! = 0 "Streamed when true." if: Boolean) on FIELD

"The `Byte` scalar type represents non-fractional whole numeric values. Byte can represent values between 0 and 255."
scalar Byte

scalar ByteArray

"The `DateTime` scalar represents an ISO-8601 compliant date time type."
scalar DateTime @specifiedBy(url: "https:\/\/www.graphql-scalars.com\/date-time")

"The built-in `Decimal` scalar type."
scalar Decimal

"The `Long` scalar type represents non-fractional signed whole 64-bit numeric values. Long can represent values between -(2^63) and 2^63 - 1."
scalar Long

"The `Short` scalar type represents non-fractional signed whole 16-bit numeric values. Short can represent values between -(2^15) and 2^15 - 1."
scalar Short

"IEEE 754 32 bit float"
scalar Single
